### PR TITLE
Feat custom enumerable methods

### PR DESCRIPTION
**Pull request by trekab**
This PR shows my endeavors to create some custom enumerable methods as required by the odin project. Project 2(https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks).

### CODE REVIEW

File: enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    my_each { |e| yield(index(e), e) }
+    self
+  end
```
Comment (ebukaume) [inline enumerable_methods.rb]: You seem to have swapped the order of the block variables here.  :point_right:  `each with index` :wink: 

`[2, 4, 6].each_with_index { |num, index| puts "#{num} -> #{index}" }` should output:

```
2 -> 0
4 -> 1
6 -> 2
```

File: enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    selected_items = []
+    my_each { |item| selected_items << item if yield(item) }
+    selected_items
+  end
+
+  def my_all?
+    my_each { |item| return false unless yield(item) }
+    true
+  end
```
Comment (ebukaume) [inline enumerable_methods.rb]: What happens when a block is not given?

`[1, true, :sym].my_all?` ==  `[1, true, :sym].all?` == `true`

`[1, 2, 3].my_all?(Integer)` ==  `[1, 2, 3].all?(Integer)` == `true`

However, you method throws an error:

> ArgumentError:
       wrong number of arguments (given 1, expected 0)

I advise you to test all your method against the once ruby provide.

File: enumerable_methods.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  def my_any?
+    my_each { |item| return true if yield(item) }
+    false
+  end
```
Comment (ebukaume) [inline enumerable_methods.rb]: What happens when no block is given? From your implementation, you always return `false`. So if I run `[nil, true].my_any?` I would get `false`. You need to fix this.

Consider having another execution path for when a block is not given.

Also, when I run `[nil, true, 2].my_any?(Integer)`, I get the error below:

> ArgumentError:
>        wrong number of arguments (given 1, expected 0)

File: enumerable_methods.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+  def my_none?
+    return true unless block_given?
+
+    my_each { |item| return true unless yield(item) }
+    false
+  end
```
Comment (ebukaume) [inline enumerable_methods.rb]: Same problem as with `#my_all?`

File: enumerable_methods.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    mapped_array
+  end
+
+  def my_inject
+    sum = 1
+    my_each do |item|
+      sum = yield(sum, item)
+    end
+    sum
+  end
```
Comment (ebukaume) [inline enumerable_methods.rb]: `[1, 2, 3].my_inject(:+)` throws an error.

>       ArgumentError:
>         wrong number of arguments (given 1, expected 0)

Also, [10, 2, 3].my_inject(&:+) returns 16 (it seems to always be 1 more than the correct answer).

**Review by ebukaume — CHANGES_REQUESTED** [review]
## Changes required! :stop_sign: 

Hi @trekab  :wave: 

Nice job so far. Some of the tests are passing but there are a few you need to fix. Kindly see below for more details.

**Kindly submit another code review after fixing these issues**. _You can always run your own tests to make sure you got it right_ :wink: 

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

File: enumerable_methods.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    my_each { |item| selected_items << item if yield(item) }
+    selected_items
+  end
+
+  def my_all?
+    return true unless block_given?
+
+    my_each { |item| return false unless yield(item) }
+    true
+  end
```
Comment (Redvanisation) [inline enumerable_methods.rb]: my_all? method still gives an error while running the code below:

`[1, 2, 3].my_all?(Integer) == [1, 2, 3].all?(Integer) == true`

> ArgumentError:
> wrong number of arguments (given 1, expected 0)

File: enumerable_methods.rb [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+  def my_any?
+    return true unless block_given?
+
+    my_each { |item| return true if yield(item) }
+    false
+  end
```
Comment (Redvanisation) [inline enumerable_methods.rb]: my_any? is still returning the error below when we run the following code:  `[nil, true, 2].my_any?(Integer)`

> ArgumentError:
> wrong number of arguments (given 1, expected 0)

File: enumerable_methods.rb [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    mapped_array
+  end
+
+  def my_inject(*args)
+    init = args.size.positive?
+    acc = init ? args[0] : self[0]
+    drop(init ? 0 : 1).my_each { |item| acc = yield(acc, item) }
+    acc
+  end
+end
```
Comment (Redvanisation) [inline enumerable_methods.rb]: When running my_inject with this code it gives the following error:

`[1, 2, 3].my_inject(:+)`

>       Failure/Error: expect(array.my_inject(:+)).to eq array.inject(:+)
>       
>       LocalJumpError:
>         no block given (yield)

**Review by Redvanisation — CHANGES_REQUESTED** [review]
# Changes required  ♻️ 

### Hello @trekab   🌟

**Good work on this project, but we need to fix the issues below before moving on ⬇️** 

- Please check my comments below and fix the mentioned errors, check [this](https://ruby-doc.org/core-2.6.5/Enumerable.html) for reference.

### _After making the required changes please submit a new code review request_

### Good luck! 🍀

**Comment by trekab** [comment]
I am done with the required changes!!!

File: enumerable_methods.rb [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    my_each { |i| return false unless i } unless block_given?
+    true
+  end
+
+  def my_any?(args = nil)
+    my_each { |num| return true if yield(num) }
+    return true unless args.nil?
+
+    false
+  end
```
Comment (imhta) [inline enumerable_methods.rb]: There are some issue with `my_any?` method
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
- when a pattern other than Regex or a Class is given returns false if none of the s are false or nil
collection matches the pattern

File: enumerable_methods.rb [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_none?
+    return true unless block_given?
+
+    my_each { |item| return true unless yield(item) }
+    false
+  end
```
Comment (imhta) [inline enumerable_methods.rb]: There are some issue with `my_none?` method
- when no block or argument is given returns true only if none of the collection mion is a member of such class    embers is true
- when a class is passed as an argument returns true if none of the collection is collection matches the pattern   
a member of such class
- when a Regex is passed as an argument returns true only if none of the collectioa member of such classn matches the Regex
- when a pattern other than Regex or a Class is given returns true only if none of the collection matches the patte the collection matches the pattern

**Review by imhta — CHANGES_REQUESTED** [review]
# Require Changes 🛑 

### Hello @trekab, 👋 `my_none` and `my_any` methods are failing the test. try to fix it. kindly check the inline comment.

---
#### Don't forgot to request another code review 
---

** Pon Muthu Selvam N**
_If you have any queries, feel free to ping me on slack_

File: enumerable_methods.rb [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    my_each { |e| yield(e, index(e)) }
+    self
+  end
```
Comment (imhta) [inline enumerable_methods.rb]: In `my_each_with_index`
- my_each_with_index calls the given block once for each element in self

File: enumerable_methods.rb [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    selected_items
+  end
+
+  def my_all?(*args)
+    return grep(args.first).length == size unless args.empty?
+
+    my_each { |i| return false unless yield(i) } if block_given?
+    my_each { |i| return false unless i } unless block_given?
+    true
+  end
```
Comment (imhta) [inline enumerable_methods.rb]: In `my_all?`
- my_all? when no block or argument is given returns true when none of the collection members are false or nil

File: enumerable_methods.rb [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+  def my_any?(*args)
+    return !grep(args.first).empty? unless args.empty?
+
+    my_each { |i| return true if yield(i) } if block_given?
+
+    my_each { |i| return true if i } unless block_given?
+
+    false
+  end
```
Comment (imhta) [inline enumerable_methods.rb]: In `my_any?`
- my_any? when no block or argument is given returns true if at least one of the collection is not false or nil

File: enumerable_methods.rb [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+  def my_none?(*args)
+    return grep(args.first).empty? unless args.empty?
+
+    my_each { |i| return false if yield(i) } if block_given?
+
+    my_each { |i| return false if i } unless block_given?
+
+    true
+  end
```
Comment (imhta) [inline enumerable_methods.rb]: In `my_none?`
- my_none? when no block or argument is given returns true only if none of the collection members is true

**Review by imhta — CHANGES_REQUESTED** [review]
# Require Changes 🛑 

### Hello @trekab, 👋  There are some methods are failing the test. try to fix it. kindly check the inline comment.

---
#### Don't forgot to request another code review 
---

** Pon Muthu Selvam N**
_If you have any queries, feel free to ping me on slack_

File: enumerable_methods.rb [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+        break if item == false
+      end
+    else
+      my_each do |item|
+        matching_items << item unless yield(item)
+        break if yield(item) == false
+      end
+    end
+    length > matching_items.length
+  end
```
Comment (ebukaume) [inline enumerable_methods.rb]: Consider when a `proc` is passed as the argument.

**Review by ebukaume — CHANGES_REQUESTED** [review]
## Changes Required! :stop_sign: 

Hi @trekab :wave: 

You are really doing great. I like your determination and perseverance, please keep it up. Let fix these 2 issues that are still there in your solution.
- [ ] `my_any?` returns a wrong value when a `proc` is passed.
- [ ] `[false, nil].my_any?` is expected to return `false` but I am getting `true`.

**Kindly submit another code review request after making the fixes**

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

**Review by ebukaume — APPROVED** [review]
We are now good to go :+1:  **You can go ahead and merge this branch**

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

**Comment by trekab** [comment]
:relaxed: :handshake:

**Pull request merged**